[feature] display logs and failure informations

Would it be possible to display the output of mocha and also why the failing tests failed?

Feel free to close, if this is already provided and I just wasn't able to run it properly. (I am thinking about the other issues we are currently fixing)

i think thats already there: (used es5-project\src\test\utility.test.js)
when i run first two test (by commenting out rest) log looks like this:
```
Grep pattern:
  ^Utility

Test file(s):
c:\projects\github\vscode-mocha-test-runner\.vscode-test\es5-project\src\test\utility.test.js

Starting mocha ...

  Utility
    √ Success (1005ms)
    1) Fail

  1 passing (1s)
  1 failing

  1) Utility
       Fail:

      AssertionError: expected 3 to equal 0
      + expected - actual

      -3
      +0
      
      at Context.<anonymous> (src\test\utility.test.js:15:36)
```
with tells u that first test succed and second fails because 3 != 0 ...

--- a/README.md
+++ b/README.md
@@ -1,4 +1,4 @@
-# Mocha test runner 
+# Mocha test runner
 
 Creates code lens before each **describe**, **it**, **suite** and **test** function call in test files.
 
@@ -6,7 +6,23 @@ Lense on **it** and **test** function shows last test result and can be used to 
 
 Lense on **describe** and **suite** shows aggregated results from nested **it** and **test** and can be used to run all tests in given group / only test with specific status.
 
-Tests not writen in javascript that mocha can load (es6 with import command, razor syntax, typescript, etc.) needs to be transpiled to es5 or es6 without import command - see example projects in .vscode-test directory. Mocha is searching for transpiled scripts ```**/*.test.js``` (configurable - **mocha.glob**) in project root directory or in **mocha.sourceDir** (appended to project root directory - used in tandem with **mocha.outputDir** for non es5 sources - see examples).
+To run tests that need to be transpiled (es6 with imports, typescript, ..) are supported using one of the two following methods:
+
+To ensure Mocha test runner works property don't forget to set the **mocha.nodeExec** variable in
+your settings. You must provide an absolute path.
+
+### Map a source directory to a build directory
+This method is useful if you want to provide your own transpile command, and want to map the resulting build directory to your source directory.
+see example projects in .vscode-test directory. Mocha is searching for transpiled scripts ```**/*.test.js``` (configurable - **mocha.glob**) in project root directory or in **mocha.sourceDir** (appended to project root directory - used in tandem with **mocha.outputDir** for non es5 sources - see examples).
+
+### Automatically transpile them
+If you want to automatically run a transpiler on your source, using the equivalent of the --require mocha parameter, set the **mocha.requires** variable in your project settings.
+
+For example you can use the following:
+
+- babel: `mocha.requires: ['babel-register']`
+- typescript: `mocha.requires: ['ts-node/register']`
+- coffeescript: `mocha.requires: ['coffee-script/register']`
 
 
 ## Preview:
@@ -24,7 +40,7 @@ Tests not writen in javascript that mocha can load (es6 with import command, raz
 
 ## Known issues:
 - (Wont fix) Tests not wrapped by describe function are ignored. (mocha is not returning file path for these - no way to unique identify tests with same title in different files)
-- Debugger wont see breakpoints in ts file when compiled with webpack (source maps issue)
+- Debugger won't see breakpoints in ts file when compiled with webpack (source maps issue)
 
 ## TODO:
 - remove "mocha.SourceDir" property and use sourcemaps to search for original source file from transpiled file.

--- a/package.json
+++ b/package.json
@@ -39,6 +39,10 @@
                     "default": {},
                     "description": "Options passed to Mocha."
                 },
+                "mocha.requires": {
+                    "default": {},
+                    "description": "Array of module required to run Mocha. (for example babel-register, coffee-script/register or ts-node/register)"
+                },
                 "mocha.env": {
                     "default": {},
                     "description": "Environment variables to run your tests."

--- a/src/TestProcess.ts
+++ b/src/TestProcess.ts
@@ -68,6 +68,7 @@ function createMocha() {
         reporterOptions?: any;
         slow?: number;
         bail?: boolean;
+        requires?: string[];
     } = undefined;
 
     if (args.options) {

--- a/src/TestProcess.ts
+++ b/src/TestProcess.ts
@@ -12,10 +12,12 @@ const suitePath: string[] = [];
 process.on('message', processArgs => {
     runTestProcess(processArgs)
         .then(results => {
+            console.log("Mocha VSCode Runner - done")
             process.send(results);
             process.exit(0);
         })
         .catch(err => {
+            console.log('Mocha VSCode Runner Error:', err)
             process.send(err);
             process.exit(-1);
         });

--- a/src/TestRunner.ts
+++ b/src/TestRunner.ts
@@ -5,9 +5,9 @@ import * as fs from 'fs';
 import { Glob } from 'glob';
 import { ForkOptions } from 'child_process';
 import { config } from "./config";
-import { fork, throwIfNot } from "./Utils";
 import { TestProcessRequest, TestsResults, TestProcessResponse } from "./Types";
 import { runTestProcess } from "./TestProcess";
+import {spawnTestProcess, throwIfNot} from './Utils';
 
 export function runTests(debug = false, grep?: string, fileSelectors?: string[]) {
     return runTestsCore({ grep, fileSelectors }, debug);
@@ -32,20 +32,19 @@ function runTestsCore(processArgs: Partial<TestProcessRequest>, debug: boolean) 
 
     const testProcess = path.join(path.dirname(module.filename), 'TestProcess.js');
 
-    const forkOptions = {
+    const spawnTestProcessOptions = {
         cwd: vscode.workspace.rootPath,
-        silent: true,
         env: config.env,
         execPath: config.nodeExec,
-        execArgv: []
+        execArgv: [],
     };
 
     if (debug) {
-        forkOptions.execArgv = ['--debug-brk=' + config.debugPort];
+        spawnTestProcessOptions.execArgv = ['--debug-brk=' + config.debugPort];
     }
-    
-    const childProcess = fork(testProcess, [], forkOptions);
-    
+
+    const childProcess = spawnTestProcess(testProcess, [], spawnTestProcessOptions);
+
     if (debug) {
         vscode.commands.executeCommand('vscode.startDebug', {
             "name": "Attach",
@@ -65,7 +64,7 @@ function runTestsCore(processArgs: Partial<TestProcessRequest>, debug: boolean) 
                 path.join(args.workspacePath, args.rootPath, "**/*.js")
             ],
         });
-        
+
         args.options = { ...args.options, timeout: 360000 };
     }
 
@@ -84,19 +83,19 @@ function runTestsCore(processArgs: Partial<TestProcessRequest>, debug: boolean) 
             results = data;
         });
 
-        childProcess.stdout.on('data', data => { 
-            if (typeof data !== 'string') { 
+        childProcess.stdout.on('data', data => {
+            if (typeof data !== 'string') {
                 data = data.toString('utf8');
             }
 
             stdout.push(data);
         });
 
-        childProcess.stderr.on('data', data => { 
-            if (typeof data !== 'string') { 
+        childProcess.stderr.on('data', data => {
+            if (typeof data !== 'string') {
                 data = data.toString('utf8');
             }
-            
+
             if (data.startsWith('Warning:') ||
                 data.startsWith('Debugger listening on')) {
                 stdout.push(data);
@@ -104,8 +103,8 @@ function runTestsCore(processArgs: Partial<TestProcessRequest>, debug: boolean) 
             }
 
             stderr.push(data);
-            if (!stderrTimeout) { 
-                stderrTimeout = setTimeout(() => { 
+            if (!stderrTimeout) {
+                stderrTimeout = setTimeout(() => {
                     childProcess.kill('SIGTERM');
                     doReject();
                 }, 500);
@@ -116,9 +115,9 @@ function runTestsCore(processArgs: Partial<TestProcessRequest>, debug: boolean) 
             if (code !== 0) {
                 if (stderrTimeout) {
                     pendingReject = true;
-                } else { 
+                } else {
                     doReject();
-                }    
+                }
             } else {
                 resolve({
                     results,
@@ -128,12 +127,12 @@ function runTestsCore(processArgs: Partial<TestProcessRequest>, debug: boolean) 
         });
 
         if (debug) {
-            // give debugger some time to properly attach itself before running tests ...        
+            // give debugger some time to properly attach itself before running tests ...
             setTimeout(() => {
                 childProcess.send(args);
             }, 1000);
         } else {
             childProcess.send(args);
-        }       
+        }
     });
 }

--- a/src/TestRunner.ts
+++ b/src/TestRunner.ts
@@ -37,6 +37,7 @@ function runTestsCore(processArgs: Partial<TestProcessRequest>, debug: boolean) 
         env: config.env,
         execPath: config.nodeExec,
         execArgv: [],
+        requires: config.requires || [],
     };
 
     if (debug) {

--- a/src/Utils.ts
+++ b/src/Utils.ts
@@ -64,11 +64,16 @@ export function spawnTestProcess(modulePath: string, args?: string[], options?: 
     execArgv = [...(options.execArgv)];
 
 
-    args = execArgv.concat([modulePath], args);
     options.stdio = ['pipe', 'pipe', 'pipe', 'ipc']
 
+    let requiresArgs = []
+    if (options.requires) {
+        for (const req of options.requires) {
+            requiresArgs.push('-r', req)
+        }
     }
 
+    args = execArgv.concat(requiresArgs, [modulePath], args);
 
     console.log('spawning:', options.execPath, args, options)
     return spawn(options.execPath, args, options);

--- a/src/Utils.ts
+++ b/src/Utils.ts
@@ -8,6 +8,7 @@ import { spawn, ForkOptions, ChildProcess } from 'child_process';
 export function getFileSelector(fileName: string) {
     throwIfNot('getFileSelector', fileName, 'fileName');
 
+
     const selector = path.relative(vscode.workspace.rootPath, fileName);
     const index = selector.lastIndexOf('.');
     return (index === -1) ? selector : selector.substring(0, index);
@@ -45,71 +46,30 @@ export function appendError(err) {
     outputChannel.appendLine(err);
 }
 
-// sigtly modified version of child_process.fork
-export function fork(modulePath: string, args?: string[], options?: ForkOptions): ChildProcess {
+export interface SpawnTestProcessOptions {
+  cwd?: string;
+  env?: any;
+  execPath?: string;
+  execArgv?: string[];
+  stdio?: any[];
+  requires?: string[];
+}
+
+// spawn our TestProcess program
+export function spawnTestProcess(modulePath: string, args?: string[], options?: SpawnTestProcessOptions): ChildProcess {
     // Get options and args arguments.
     var execArgv: string[];
 
-    // COMMENTED OUT: (using typescript optional params instead)
-    // var options: any = {};
-    // var args: any = [];
-    // var pos = 1;
-    // if (pos < arguments.length && Array.isArray(arguments[pos])) {
-    //     args = arguments[pos++];
-    // }
-
-    // if (pos < arguments.length && arguments[pos] != null) {
-    //     if (typeof arguments[pos] !== 'object') {
-    //         throw new TypeError('Incorrect value of args option');
-    //     }
-
-    //     options = util._extend({}, arguments[pos++]);
-    // }
-
     // Prepare arguments for fork:
-    execArgv = [...(options.execArgv || process.execArgv)];
+    execArgv = [...(options.execArgv)];
 
-    if (execArgv === process.execArgv && (process as any)._eval != null) {
-        const index = execArgv.lastIndexOf((process as any)._eval);
-        if (index > 0) {
-            // Remove the -e switch to avoid fork bombing ourselves.
-            execArgv = execArgv.slice();
-            execArgv.splice(index - 1, 2);
-        }
-    }
-
-    // ADDED: remove --inspect-brk= atribute from original arguments (in case of debugging extension)
-    const index = execArgv.findIndex(o => o.startsWith('--inspect-brk='));
-    if (index !== -1) {
-        execArgv.splice(index, 1);
-    }
 
     args = execArgv.concat([modulePath], args);
+    options.stdio = ['pipe', 'pipe', 'pipe', 'ipc']
 
-    if (!Array.isArray(options.stdio)) {
-        // Use a separate fd=3 for the IPC channel. Inherit stdin, stdout,
-        // and stderr from the parent if silent isn't set.
-        options.stdio = options.silent ? ['pipe', 'pipe', 'pipe', 'ipc'] :
-            [0, 1, 2, 'ipc'];
-    } else if (options.stdio.indexOf('ipc') === -1) {
-        throw new TypeError('Forked processes must have an IPC channel');
     }
 
-    // When forking a child script, we setup a special environment to make
-    // the atom-shell binary run like the upstream node.
-    if (!options.env) {
-        options.env = Object.create(process.env);
-    }
 
-    // REMOVED - dont wanna electorn in the mix
-    //options.env.ELECTRON_RUN_AS_NODE = 1;
-
-    // On Mac we use the helper app as node binary.
-    if (!options.execPath && (process as any).type && process.platform == 'darwin') {
-        options.execPath = (process as any).helperExecPath;
-    }
-
-    options.execPath = options.execPath || process.execPath;
-
+    console.log('spawning:', options.execPath, args, options)
     return spawn(options.execPath, args, options);
 }

--- a/src/config.ts
+++ b/src/config.ts
@@ -13,6 +13,7 @@ interface Config {
     debugTrace: string;
     compilerScript: string;
     nodeExec: string;
+    requires: string[];
 
     update: (section: string, value: any) => void;
 }
@@ -32,6 +33,7 @@ export const config: Config = {
     get debugTrace() { return get<string>('debugTrace'); },
     get compilerScript() { return get<string>('compilerScript'); },
     get nodeExec() { return get<string>('nodeExec'); },
+    get requires() { return get<string[]>('requires'); },
     update: (section: string, value: any) => {
         vscode.workspace.getConfiguration('mocha').update(section, value);
     }

